archive/imo/imo1969_q1.lean

```diff
@@ -0,0 +1,81 @@
+import tactic
+import tactic.basic
+import tactic.linarith
+import tactic.norm_cast
+import tactic.ring
+
+open int
+open nat
+
+/-
+The 1969 IMO, problem 1:
+Prove that there are infinitely many natural numbers $a$ with the following property:
+the number $z = n^4 + a$ is not prime for any natural number $n$.
+
+The key to the solution is that you can factor this into the product of two polynomials, if a = 4*m^4.
+-/
+
+lemma factorization (m n: ℤ): (n^2 + 2*m^2 - 2*n*m) * (n^2 + 2*m^2 + 2*n*m) = n^4 + 4*m^4
+:= by ring
+
+/-
+To show that the product is not prime, we need to show each of the factors is at least 2, which we can
+do with a sum-of-squares expression.
+-/
+lemma left_factor_large (m n: ℤ) (h: m > 1): (n^2 + 2*m^2 - 2*n*m) > 1 :=
+have h: (n^2 + 2*m^2 - 2*n*m) = (m-n)^2 + m^2, by ring,
+begin
+  rw h,
+  nlinarith,
+end
+
+lemma right_factor_large (m n: ℤ) (h: m > 1): (n^2 + 2*m^2 + 2*n*m) > 1 :=
+have h: (n^2 + 2*m^2 + 2*n*m) = (m+n)^2 + m^2, by ring,
+begin
+  rw h,
+  nlinarith,
+end
+
+/-
+The factorization is over the integers, but we need the nonprimality over the natural numbers.
+-/
+
+lemma int_large (a: ℤ) (h: a > 1) : a.nat_abs > 1 :=
+have a = ↑(a.nat_abs) ∨ a = -↑(a.nat_abs), from nat_abs_eq a,
+or.elim this
+  (assume: a = ↑(a.nat_abs), by linarith)
+  (assume: a = -↑(a.nat_abs), by linarith)
```

Comment:
```suggestion
lemma int_large (a: ℤ) (h: 1 < a) : 1 < a.nat_abs :=
or.elim (nat_abs_eq a) (λ _, by linarith) (λ _, by linarith)
```
-- less legible but the lemma is trivial anyway.

Reply:
I'd rather keep this more legible, if you don't mind. The lemma is trivial from a human mathematician point of view, but I didn't find it trivial to figure out a reasonable way to formalize it, so the code is still representing some nontrivial thought process and it is useful for it to be legible.

Reply:
Another version is

```lean
lemma int_large (a : ℤ) (h : 1 < a) : 1 < a.nat_abs :=
by exact_mod_cast lt_of_lt_of_le h le_nat_abs
```

using the `le_nat_abs` lemma rather than splitting into cases.

Reply:
Ok, I'll use this.